File: Forge/src/main/java/ram/talia/moreiotas/forge/eventhandlers/ChatEventHandler.java
package ram.talia.moreiotas.forge.eventhandlers;

import net.minecraft.world.entity.player.Player;
import net.minecraftforge.event.ServerChatEvent;
import net.minecraftforge.eventbus.api.SubscribeEvent;
import org.jetbrains.annotations.Nullable;

import java.util.HashMap;
import java.util.Map;
import java.util.UUID;

public class ChatEventHandler {
    private static final String TAG_CHAT_PREFIX = "moreiotas:prefix";

    private static final Map<UUID, @Nullable String> lastMessages = new HashMap<>();

    private static @Nullable String lastMessage = null;

    public static void setPrefix(Player player, @Nullable String prefix) {
        if (prefix == null)
            player.getPersistentData().remove(TAG_CHAT_PREFIX);
        else
            player.getPersistentData().putString(TAG_CHAT_PREFIX, prefix);
    }

    public static @Nullable String getPrefix(Player player) {
        if (!player.getPersistentData().contains(TAG_CHAT_PREFIX))
            return null;
        return player.getPersistentData().getString(TAG_CHAT_PREFIX);
    }

    public static @Nullable String getLastMessage(@Nullable Player player) {
        if (player == null)
            return lastMessage;
        return lastMessages.get(player.getUUID());
    }

    @SubscribeEvent
    public static void chatMessageSent(ServerChatEvent event) {
        var player = event.getPlayer();
        var uuid = player.getUUID();
        var text = event.getRawText();

        if (event.isCanceled())
            return;

        if (!player.getPersistentData().contains(TAG_CHAT_PREFIX)) {
            lastMessages.put(uuid, text);
            lastMessage = text;
            return;
        }

        var prefix = player.getPersistentData().getString(TAG_CHAT_PREFIX);

        if (text.startsWith(prefix)) {
            event.setCanceled(true);
            lastMessages.put(uuid, text.substring(prefix.length()));
            return;
        }

        lastMessage = text;
    }
}

File: Fabric/src/main/java/ram/talia/moreiotas/fabric/eventhandlers/ChatEventHandler.kt
package ram.talia.moreiotas.fabric.eventhandlers

import net.minecraft.network.chat.ChatType
import net.minecraft.network.chat.PlayerChatMessage
import net.minecraft.server.level.ServerPlayer
import net.minecraft.world.entity.player.Player
import ram.talia.moreiotas.fabric.cc.MoreIotasCardinalComponents
import java.util.*

object ChatEventHandler {
    private val lastMessages: MutableMap<UUID, String?> = mutableMapOf()
    private var lastMessage: String? = null

    fun receiveChat(message: PlayerChatMessage, player: ServerPlayer, params: ChatType.Bound): Boolean {
        val text = message.signedBody.content + (message.unsignedContent?.string ?: "")

        val prefix = MoreIotasCardinalComponents.CHAT_PREFIX_HOLDER.get(player).prefix

        if (prefix == null) {
            lastMessages[player.uuid] = text
            lastMessage = text
            return true
        }

        if (!text.startsWith(prefix))
            return true

        lastMessages[player.uuid] = text.substring(prefix.length)

        return false
    }

    @JvmStatic
    fun lastMessage(player: Player?): String? = if (player != null) lastMessages[player.uuid] else lastMessage
}